shell/platform/darwin/ios/platform_view_ios.mm

```diff
@@ -17,39 +17,78 @@
-                                 FlutterView* owner_view)
-    : HeadlessPlatformViewIOS(delegate, std::move(task_runners)),
-      owner_controller_(owner_controller),
-      owner_view_(owner_view),
-      ios_surface_(owner_view_.createSurface) {
-  FML_DCHECK(ios_surface_ != nullptr);
-  FML_DCHECK(owner_controller_ != nullptr);
-  FML_DCHECK(owner_view_ != nullptr);
-}
+PlatformViewIOS::PlatformViewIOS(PlatformView::Delegate& delegate, blink::TaskRunners task_runners)
+    : PlatformView(delegate, std::move(task_runners)) {}
 
 PlatformViewIOS::~PlatformViewIOS() = default;
 
-FlutterViewController* PlatformViewIOS::GetOwnerViewController() const {
+PlatformMessageRouter& PlatformViewIOS::GetPlatformMessageRouter() {
+  return platform_message_router_;
+}
+
+// |shell::PlatformView|
+void PlatformViewIOS::HandlePlatformMessage(fml::RefPtr<blink::PlatformMessage> message) {
+  platform_message_router_.HandlePlatformMessage(std::move(message));
+}
+
+fml::WeakPtr<FlutterViewController> PlatformViewIOS::GetOwnerViewController() const {
   return owner_controller_;
 }
 
+void PlatformViewIOS::SetOwnerViewController(fml::WeakPtr<FlutterViewController> owner_controller) {
+  if (ios_surface_ || !owner_controller) {
+    NotifyDestroyed();
+    ios_surface_.reset();
+    accessibility_bridge_.reset();
+  }
+  owner_controller_ = owner_controller;
+  if (owner_controller_) {
+    ios_surface_ = static_cast<FlutterView*>(owner_controller.get().view).createSurface;
+    FML_DCHECK(ios_surface_ != nullptr);
+
+    if (accessibility_bridge_) {
+      accessibility_bridge_.reset(
+          new AccessibilityBridge(static_cast<FlutterView*>(owner_controller_.get().view), this));
+    }
+    // Do not call `NotifyCreated()` here - let FlutterViewController take care
+    // of that when its Viewport is sized.  If `NotifyCreated()` is called here,
+    // it can occasionally get invoked before the viewport is sized resulting in
+    // a framebuffer that will not be able to completely attach.
+  }
+}
+
 void PlatformViewIOS::RegisterExternalTexture(int64_t texture_id,
                                               NSObject<FlutterTexture>* texture) {
   RegisterTexture(std::make_shared<IOSExternalTextureGL>(texture_id, texture));
 }
 
 // |shell::PlatformView|
 std::unique_ptr<Surface> PlatformViewIOS::CreateRenderingSurface() {
-  return ios_surface_->CreateGPUSurface();
+  if (!ios_surface_) {
+    FML_DLOG(INFO) << "Could not CreateRenderingSurface, this PlatformViewIOS "
+                      "has no ViewController.";
+    return nullptr;
+  }
+
+  fml::AutoResetWaitableEvent latch;
+  std::unique_ptr<Surface> surface;
+  auto gpu_task = [ios_surface = ios_surface_.get(), &surface, &latch]() mutable {
+    if (ios_surface) {
+      surface = ios_surface->CreateGPUSurface();
+    }
+    latch.Signal();
+  };
+
+  task_runners_.GetGPUTaskRunner()->PostTask(gpu_task);
+  latch.Wait();
```

Comment:
This can deadlock if the GPU and platform task runners are the same. If you ever need to introduce a latch, please use `RunNowOrPostTask` instead. But, in this particular instance, I think acquiring the render surface in the in response to the notify created callback and stashing it in an ivar before returning it here works better. This method can then just say `return stashed_ios_surface` (or whatever you name your ivar.

Reply:
The surface has to be created on the GPU thread though, otherwise bad things can happen.  If I do it in the notify created callback, I still need to run a task on the GPU thread.  This was a bug in the existing embedding, but it doesn't really get exercised in a normal app with one viewcontroller/shell that lives and dies with the process.

It sounds like what I need here is just `RunNowOrPostTask` - or am I missing something?

Reply:
I'm not able to reproduce whatever caused an issue here - it's entirely possible the issue was due to a but in the refactoring that I've since fixed.  For now, I'll get rid of the latch/post task entirely.